common/searchattribute/name_type_map.go

```diff
@@ -33,7 +34,16 @@ func buildIndexNameTypeMap(indexSearchAttributes map[string]*persistencespb.Inde
 	return indexNameTypeMap
 }
 
+// NewNameTypeMap creates a new NameTypeMap with the given custom search attributes.
+func NewNameTypeMap(customSearchAttributes map[string]enumspb.IndexedValueType) NameTypeMap {
+	return NameTypeMap{
+		customSearchAttributes: customSearchAttributes,
+	}
+}
+
 func (m NameTypeMap) System() map[string]enumspb.IndexedValueType {
+	system := defs.System()
+	predefined := defs.Predefined()
```

Comment:
Maybe you should define `system` (and others) globally in this package as well to avoid copying in every call.